cumulusci/utils/yaml/cumulusci_yml.py

```diff
@@ -1,4 +1,264 @@
+    preflight_message: str = None
+
+
+class Project(CCIDictModel):
+    name: str = None
+    package: Package = None
+    test: Test = None
+    git: Git = None
+    dependencies: List[Dict[str, URL]] = None  # TODO
+    source_format: Literal["sfdx", "mdapi"] = "mdapi"
+
+
+class ScratchOrg(CCIDictModel):
+    config_file: Path
+    days: int = None
+    namespaced: str = None
+    setup_flow: str = None
+
+
+class Orgs(CCIDictModel):
+    scratch: Dict[str, ScratchOrg]
+
+
+class ServiceAttribute(CCIDictModel):
+    description: str
+    required: bool
+
+
+class Service(CCIDictModel):
+    description: str
+    attributes: Dict[str, ServiceAttribute]
+    validator: PythonClassPath = None
+
+
+class CumulusCIConfig(CCIDictModel):
+    keychain: PythonClassPath
+
+
+class Source(CCIDictModel):
+    github: URL = None
+    release: Literal["latest", "previous", "latest_beta"] = None
+    ref: str = None
+    branch: str = None
+    tag: str = None
+
+
+class CumulusCLIConfig(CCIDictModel):
+    show_stacktraces: bool = False
+    plain_output: bool = None
+
+
+class CumulusCIRoot(CCIDictModel):
+    tasks: Dict[str, Task] = {}
+    flows: Dict[str, Flow] = {}
+    project: Project = {}
+    orgs: Orgs = {}
+    services: Dict[str, Service] = {}
+    cumulusci: CumulusCIConfig = None
+    plans: Dict[str, Plan] = []
+    minimum_cumulusci_version: str = None
+    sources: Dict[str, Source] = []
+    cli: CumulusCLIConfig = None
+
+
+class CumulusCIFile(CCIDictModel):
+    __root__: Union[CumulusCIRoot, None]
+
+
+def parse_from_yaml(source):
+    "Parse from a path, url, path-like or file-like"
+    return CumulusCIFile.parse_from_yaml(source)
+
+
+def validate_data(
+    data: Union[dict, list],
+    context: str = None,
+    on_error: callable = None,
+):
+    """Validate data which has already been loaded into a dictionary or list.
+
+    context is a string that will be used to give context to error messages.
+    on_error will be called for any validation errors with a dictionary in Pydantic error format
+
+    https://pydantic-docs.helpmanual.io/usage/models/#error-handling
+    """
+    return CumulusCIFile.validate_data(data, context=context, on_error=on_error)
+
+
+class ErrorDict(TypedDict):
+    "The structure of a Pydantic error dictionary. Google TypedDict if its new to you."
+    loc: Sequence[Union[str, int]]
+    msg: str
+    type: str
+
+
+def _log_yaml_error(logger, error: ErrorDict):
+    "Format and log a Pydantic-style error dictionary"
+    logger.warning("CumulusCI Parsing Error:")
+    loc = " -> ".join(repr(x) for x in error["loc"] if x != "__root__")
+    logger.warning("%s : %s", loc, error["msg"])
+    logger.error(
+        "NOTE: These errors will cause major problems in future versions of CumulusCI."
+    )
+    logger.error(
+        "If you think your YAML has no error, please report the bug to the CumulusCI team."
+    )
+    logger.error("https://github.com/SFDO-Tooling/CumulusCI/issues/")
```

Comment:
The output is a little hard to read:

<img width="1044" alt="Screen Shot 2021-04-07 at 11 40 39 AM" src="https://user-images.githubusercontent.com/49014/113894658-38470880-9796-11eb-9c69-7f7441d64cb9.png">

Suggestions:
- Only show "CumulusCI Parsing Error", the note, and the link to the issue tracker once
- Put a newline and indentation between the location and error message
- Log an empty line after all the warnings

There are a couple other issues here too:
- "Steps must have a task or flow" shouldn't be shown here; the step in question has a task or flow inherited from the universal config
- `checks` is allowed in a flow step

Reply:
Overriding the options without specifying what task you are overriding seems quite fragile to me. Do we want to encourage it?

Reply:
I don't think we've encouraged it, but it's possible and clearly done.